File: examples/threads/SynchronizedPingPong.java

/**
 * Creates two threads that play ping-pong in a synchronized manner using
 * wait/notify methods.
 * 
 * @author amit
 */

public class SynchronizedPingPong
{
    public static void main(String[] args)
    {
	if (args.length != 1) {
	    System.err.println("Usage: java SynchronizedPingPong <delay>");
	    System.exit(1);
	}
	int delay = Integer.parseInt(args[0]);
	Ping ping = new Ping("ping", delay);
	Pong pong = new Pong("PONG", delay);
	ping.setPong(pong);
	pong.setPing(ping);

	new Thread(ping).start();
	new Thread(pong).start();
    }
}

class Ping implements Runnable
{
    private String word; // what word to print
    private int delay; // how long to pause (in milliseconds)
    private Pong pong;

    public Ping(String whatToSay, int delayTime)
    {
	word = whatToSay;
	delay = delayTime;
    }

    public void setPong(Pong pong)
    {
	this.pong = pong;
    }

    public void run()
    {
	try {
	    for (;;) {
		System.out.println(word + " ");
		Thread.sleep(delay); // wait until next time
		synchronized (pong) {
		    pong.notify();
		}
		synchronized (this) {
		    wait();
		}
	    }
	} catch (InterruptedException e) {
	    System.err.println(e);
	    return; // end this thread
	}
    }
}

class Pong implements Runnable
{
    private String word; // what word to print
    private int delay; // how long to pause, in milliseconds
    private Ping ping;

    public Pong(String whatToSay, int delayTime)
    {
	word = whatToSay;
	delay = delayTime;
    }

    public void setPing(Ping ping)
    {
	this.ping = ping;
    }

    public void run()
    {
	try {
	    for (;;) {
		synchronized (this) {
		    wait();
		}
		System.out.println(word + " ");
		Thread.sleep(delay); // wait until next time
		synchronized (ping) {
		    ping.notify();
		}
	    }
	} catch (InterruptedException e) {
	    System.err.println(e);
	    return; // end this thread
	}
    }
}

File: examples/go/threads/SynchronizedPingPong.go
//Author: Amit Jain & Paul Ellis

package main

import (
	"fmt"
	"sync"
	"time"
	"strconv"
	"os"
)

type Ping struct{
	word string
	delay time.Duration
	ping *chan bool
	pong *chan bool
}

type Pong struct{
	word string
	delay time.Duration
	ping *chan bool
	pong *chan bool
}

func NewPing(word string, delay int, ping *chan bool, pong *chan bool) *Ping{
	//Here we need to convert an integer to a time.Delay type for time.Sleep
	t,_ := time.ParseDuration(strconv.Itoa(delay) + "ms")
	p := Ping{word:word, delay:t, ping:ping, pong:pong}
	return &p
}

func NewPong(word string, delay int, ping *chan bool, pong *chan bool) *Pong{
	//Here we need to convert an integer to a time.Delay type for time.Sleep
	t,_ := time.ParseDuration(strconv.Itoa(delay) + "ms")
	p := Pong{word:word, delay:t, ping:ping, pong:pong}
	return &p
}

func (p *Ping) Run(){
	for {
		//wait for pong to send something to the Pong channel
		//The channel will block until it gets a value
		<- *p.pong
		fmt.Println(p.word + " ")
		time.Sleep(p.delay)
		//let pong know ping is done
		*p.ping <- true
	}
}

func (p *Pong)Run(){
	for {
		//wait for ping to send something to the ping channel
		//The channel will block until it gets a value
		<- *p.ping
		fmt.Println(p.word + " ")
		time.Sleep(p.delay)
		//let ping know pong is done
		*p.pong <- true
	}
}

func main(){
	args := os.Args[1:]

	if len(args) != 1 {
		fmt.Println("Usage: $ go run SynchronizedPingPong.go <delay(ms)>")
		os.Exit(1)
	}

	//Get the delay as an integer, the underscore means that we don't
	//care if strconv.Atoi returns an error, just ignore that return val
	delay, _ := strconv.Atoi(args[0])

	//Here we are making channels that different go routines can talk
	//to each other with. A channel will block until it can recieve or
	//Send which makes it useful for this situation.
	pingChan := make(chan bool)
	pongChan := make(chan bool)

	var wg sync.WaitGroup

	ping := NewPing("ping", delay, &pingChan, &pongChan)
	pong := NewPong("pong", delay, &pingChan, &pongChan)

	//You have to give one channel a "kickstart" from another go routine
	go func(){pingChan <- true}()

	//Run infinitely as we never call wg.Done()
	wg.Add(1)
	go ping.Run()
	go pong.Run()
	wg.Wait()
}